proc/coredump.c

```diff
@@ -0,0 +1,518 @@
+			SIZE_COREDUMP_GREGSET +
+			SIZE_COREDUMP_THREADAUX;
+	const size_t NOTES_SIZE = SIZE_COREDUMP_GENAUX + threadCnt * (THREAD_NOTES_SIZE);
+	size_t currentOffset;
+	size_t stackSize;
+	void *userSp;
+	map_entry_t *e;
+	size_t i;
+
+	/* Notes */
+	currentOffset = sizeof(Elf_Ehdr) + sizeof(Elf_Phdr) * (1 + segCnt);
+	coredump_dumpPhdr(PT_NOTE, currentOffset, 0, NOTES_SIZE, 0, state);
+	currentOffset += NOTES_SIZE;
+
+	/* Memory */
+#ifdef COREDUMP_MEM_ALL
+	proc_lockSet(&process->mapp->lock);
+	e = lib_treeof(map_entry_t, linkage, lib_rbMinimum(process->mapp->tree.root));
+	while (e != NULL) {
+		if ((e->prot & PROT_READ) && (e->prot & PROT_WRITE)) {
+			coredump_dumpPhdr(PT_LOAD, currentOffset, e->vaddr, e->size, e->flags, state);
+			currentOffset += e->size;
+		}
+		e = lib_treeof(map_entry_t, linkage, lib_rbNext(&e->linkage));
+	}
+	proc_lockClear(&process->mapp->lock);
+#else
+	userSp = hal_cpuGetUserSP(ctx);
+	stackSize = coredump_stackSize(userSp, process);
+	coredump_dumpPhdr(PT_LOAD, currentOffset, userSp, stackSize, PROT_READ | PROT_WRITE, state);
+	currentOffset += stackSize;
+
+#ifdef COREDUMP_MEM_ALL_STACKS
+	for (i = 1; i < threadCnt; i++) {
+		userSp = hal_cpuGetUserSP(threadInfo[i].userContext);
+		stackSize = coredump_stackSize(userSp, process);
+		coredump_dumpPhdr(PT_LOAD, currentOffset, userSp, stackSize, PROT_READ | PROT_WRITE, state);
+		currentOffset += stackSize;
+	}
+#endif
+
+#endif /* COREDUMP_MEM_ALL */
+}
+
+
+static void coredump_dumpAllMemory(process_t *process, coredump_state_t *state)
+{
+	map_entry_t *e;
+
+	proc_lockSet(&process->mapp->lock);
+	e = lib_treeof(map_entry_t, linkage, lib_rbMinimum(process->mapp->tree.root));
+	while (e != NULL) {
+		if ((e->prot & PROT_READ) && (e->prot & PROT_WRITE)) {
+			coredump_encodeChunk(state, e->vaddr, e->size);
+		}
+		e = lib_treeof(map_entry_t, linkage, lib_rbNext(&e->linkage));
+	}
+	proc_lockClear(&process->mapp->lock);
+}
+
+static size_t coredump_segmentCount(process_t *process)
+{
+	map_entry_t *e;
+	size_t segCnt = 0;
+
+	proc_lockSet(&process->mapp->lock);
+
+	e = lib_treeof(map_entry_t, linkage, lib_rbMinimum(process->mapp->tree.root));
+	while (e != NULL) {
+		if ((e->prot & PROT_READ) && (e->prot & PROT_WRITE)) {
+			segCnt++;
+		}
+		e = lib_treeof(map_entry_t, linkage, lib_rbNext(&e->linkage));
+	}
+
+	proc_lockClear(&process->mapp->lock);
+	return segCnt;
+}
+
+#endif /* COREDUMP */
+
+void coredump_dump(unsigned int n, exc_context_t *ctx)
+{
+#ifndef COREDUMP
+	return;
+#else
+
+	char buff[CORE_BUF_SIZE_MAX];
+	coredump_threadinfo_t threadInfo[COREDUMP_THREADS_NUM];
+	size_t segCnt;
+	size_t threadCnt;
+	size_t i;
+	coredump_state_t state;
+	process_t *process;
+	cpu_context_t *cctx = hal_excToCpuCtx(ctx);
+
+	process = proc_current()->process;
+
+	/*
+	Ensure for dumped process that:
+	- saved context is coherent with stack memory,
+	- thread count, section count are fixed,
+	while the rest of the processes can run freely.
+	*/
+	proc_freeze(process);
```

Comment:
As we discussed - freezing shouldn't be really needed, as we have to kill the whole process anyway

Reply:
Ack